### PR TITLE
fix: use path prefix in navigate/replace/push calls

### PR DESCRIPTION
**Pull request by DSchau**
This PR fixes the behavior re: pathPrefix identified in #8155.

Essentially:

- Uses `withPrefix` helper in all programmatic navigation
- Removes manual prefix in onClick handler in `Link`
- Adds a number of unit tests and e2e tests

### CODE REVIEW

**Comment by pieh** [comment]
do `a` elements in static html builds have `href`s with path prefix with those changes?

**Comment by DSchau** [comment]
@pieh yup, it should! I originally had a variant where they didn't, but they do!

I validated with gatsby-dev-cli. Basically the behavior is unchanged, we'll pass through `prefixedTo` to `reach-router/link`, which will just create an `a` tag with that prefixed link, e.g. [these lines](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-link/src/index.js#L112-L116)

screen shot 2018-09-18 at 3 04 29 pm

File: integration-tests/path-prefix/cypress/integration/navigate.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/* global cy */
+const { pathPrefix } = require(`../../gatsby-config`)
+
+const withTrailingSlash = url => `${url}/`
+
+describe('navigate', () => {
+  beforeEach(() => {
+    cy.visit(`/`)
```
Comment (DSchau) [inline integration-tests/path-prefix/cypress/integration/navigate.js]: As info, `waitForRouteChange()` wasn't working here. Will give it another shot in a bit just to validate I'm not crazy.

Comment (pieh) [reply integration-tests/path-prefix/cypress/integration/navigate.js]: was it timing out? I've tried some more involved stuff and didn't work for me today too but didn't dig into that yet

**Pull request merged**